### PR TITLE
✨(forum) sort discussion topics

### LINKED ISSUE #66: Allow sorting discussion topics
## Feature Request

**Is your feature request related to a problem or unsupported use case? Please describe.**
On the view listing discussion topics, we want to allow sorting topics by each column : number of answers, number of views and date.

topics

**Describe the solution you'd like**
I propose to follow the way it is done in Django's admin using a querystring parameter.

Make the column headers clickable. When a column header is clicked, the value in this column is added as first sorting criteria and 2 buttons are added to the column header: remove from sorting and toggle sorting.

The sorting critera are set in the querystring (Django admin uses the "o" parameter) as a dot-separated list of the ids of columns that are active for sorting, in order of priority (last clicked first). Each column id is preceded by a minus sign if it should be sorted by descending order. 

For example a querystring parameter `o=6.-7` will sort first by column id 6 in ascending order, then by column 7 in descending order.

A click on "remove sorting" removes this column from sorting priorities
A click on "toggle sorting" switched between ascending and descending order.

**Describe alternatives you've considered**
In fact, this should probably be proposed as a PR to https://github.com/ellmetha/django-machina

**Discovery, Documentation, Adoption, Migration Strategy**
While preparing your pull request, follow guidelines in our [handbook](https://openfun.gitbooks.io/handbook/content/) and look at examples of pull requests that have already been merged to see how we do it.

### PR DESCRIPTION
**Pull request by carofun**
## Purpose

The view listing the different topics in the forum is by default ordered by last messages which enables to see recent activity. If we want to see which subjects in the past were dynamic, it’s interesting to be able to sort discussions from other parameters such as number of replies or views. On the other hand, if we have a lot of history, we still want to focus on the latest posts, so it’s interesting to combine sorting fields.

## Proposal

Make the column headers clickable. When a column header is clicked, the value in this column is added as first sorting criteria and 2 buttons are added to the column header: remove from sorting and toggle sorting.

The sorting critera are set in the querystring (Django admin uses the "o" parameter) as a dot-separated list of the ids of columns that are active for sorting, in order of priority (last clicked first). Each column id is preceded by a minus sign if it should be sorted by descending order.

For example a querystring parameter o=6.-7 will sort first by column id 6 in ascending order, then by column 7 in descending order.

A click on "remove sorting" removes this column from sorting priorities
A click on "toggle sorting" switched between ascending and descending order.

By default
Capture d’écran 2021-03-04 à 17 39 56

Capture d’écran 2021-03-04 à 17 40 06

Capture d’écran 2021-03-04 à 17 40 14

Ordering on two different columns
Capture d’écran 2021-03-04 à 17 40 53

changing the order of priority 2
Capture d’écran 2021-03-04 à 17 41 06

ordering on three different columns
Capture d’écran 2021-03-04 à 17 41 29

Resolve #66

### CODE REVIEW

**Comment by sampaccoud** [comment]
@carofun Nice feature :star_struck: Just commenting on the snapshots: maybe it would be clearer to add a "title" column header. I'm not sure where the "DISCUSSION" header would go... maybe above the table?

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
 
+### Added
+
+ - sort discussion topics
```
Comment (sampaccoud) [inline CHANGELOG.md:13]: Maybe give more explanations?
```suggestion
 - Allow sorting discussion topics by title, number of responses, number of views or date of latest response.
```

File: sandbox/settings.py [hunk 2]
```diff
@@ -254,9 +254,6 @@ class Base(Configuration):
     MACHINA_MARKUP_WIDGET = "ashley.editor.widgets.DraftEditor"
     MACHINA_PROFILE_AVATARS_ENABLED = False
     MACHINA_USER_DISPLAY_NAME_METHOD = "get_public_username_with_default"
-    MIGRATION_MODULES = {
```
Comment (sampaccoud) [inline sandbox/settings.py:257]: Ah! I was wondering why you added this, now I'm wondering why you're removing it. Can you explain? :stuck_out_tongue:

File: src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+            name="views_count",
+            field=models.PositiveIntegerField(
+                blank=True,
+                db_index=True,
+                default=0,
+                editable=False,
+                verbose_name="Views count",
+            ),
+        ),
+    ]
```
Comment (sampaccoud) [inline src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py:35]: I understand that you are adding this migration to add an index on these fields.
I'm a bit scared this will complicate our life if django-machina adds migrations. Maybe this militates for trying to sell it to them first and trying to integrate it in django-machina? If it doesn't work we can always come back to this PR?

**Comment by carofun** [comment]
> @carofun Nice feature 🤩 Just commenting on the snapshots: maybe it would be clearer to add a "title" column header. I'm not sure where the "DISCUSSION" header would go... maybe above the table?

Hi, !! 
Using it I find it quiet obvious the way it is now. But I might not be objective lol as I had trouble placing number and arrow and it had been messy for a while :-). I'm not sure to understand what you suggest as an improvement. Do you mean to add a label Title ? sorry about that for my misunderstanding

File: sandbox/settings.py (hunk 2 above)
Comment (carofun) [reply sandbox/settings.py:257]: no problem ! but do you know I'm not finished yet (lol I'm in wip), I actually have lint problem with my overload that I've just managed to fix, it's not released yet ! Anyway it's always a pleasure to exchange so It doesn't matter !
To answer, this was my first choice on overloading the forum_conversation app. If you don't want to copy all the migrations files you can set this option in the settings. I didn't like importing all the files knowing I had done no change in the model, I thought it could be confusing so using this option allowed that. 
For this commit, as I'm changing the model on adding index, I can't use it anymore and I have to copy all the migration files.

File: src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py (hunk 3 above)
Comment (carofun) [reply src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py:35]: In the real PR for machina I won't have to overload anything, I can directly add in the model the index so it won't look so massive. I will have only a file for the makemigration or if you want to I can even not add this modification on the PR for machina but for our project with the amount of forums and expected posts that we have and classes, it's a point that needs to be optimized IMO and as it's already done, I'm not certain to understand why it shouldn't be used

File: sandbox/settings.py (hunk 2 above)
Comment (madmatah) [reply sandbox/settings.py:257]: @sampaccoud For reference, you'll find explanations [here](https://django-machina.readthedocs.io/en/stable/customization/overriding_applications.html#import-the-application-models-if-needed) about the 2 options available concerning the migration files, when overriding a django-machina app.

File: src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py (hunk 3 above)
Comment (madmatah) [reply src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py:35]: For other django-machina apps that we have overloaded, we copied the migration files.
We just have to run the `makemigrations` command to stay up-to date with django-machina migrations, right ? (except if they write migrations manually).

File: tests/ashley/machina_extensions/forum/test_forum_view.py [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+    TopicFactory,
+    TopicSortHeaderFactory,
+    UserFactory,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class TestForumView(TestCase):
+    """ Displays a forum and its topics overridden from django machina  """
```
Comment (madmatah) [inline tests/ashley/machina_extensions/forum/test_forum_view.py:18]:
```suggestion
    """Displays a forum and its topics overridden from django machina."""
```

File: tests/ashley/machina_extensions/forum/test_forum_view.py [hunk 5]
```diff
@@ -0,0 +1,411 @@
...
+                    {
+                        "url_primary": "?o=2",
+                    },
+                    {
+                        "url_primary": "?o=3",
+                    },
+                ],
+            ),
+        )
+
```
Comment (madmatah) [inline tests/ashley/machina_extensions/forum/test_forum_view.py:101]: From what I understand, you are testing the context injected into the template.
Why not testing the HTML sent in the HTTP response ?

I understand that it's easier to test this way, but I'm worried because it implies that we consider that the templates can't contain errors.

(My comment applies to all tests using the response context)

File: tests/ashley/machina_extensions/forum/test_forum_view.py [hunk 6]
```diff
@@ -0,0 +1,411 @@
...
+        assert response.status_code == 200
+
+        # Should have 1 sorted column as the other two are unknown format
+        self.assertContains(response, "sorted ", count=1)
+
+        # it should be in alphabetic order
+        self.assertContentBefore(response, "TOPIC A", "TOPIC B")
+        self.assertContentBefore(response, "TOPIC B", "TOPIC C")
+
+    def test_sorting_context_wisiwig(self):
```
Comment (madmatah) [inline tests/ashley/machina_extensions/forum/test_forum_view.py:328]: I don't understand the concept of WYSIWYG in this context ^^

```suggestion
    def test_sorting_context_wysiwyg(self):
```

File: tests/ashley/machina_extensions/forum/test_forum_view.py [hunk 7]
```diff
@@ -0,0 +1,411 @@
...
+            (
+                '<a href="?o=-0.-2.3">Topics </a>'
+                '<span class="sortoptions">'
+                '<a class="sortremove" href="?o=-2.3" title="Remove from sorting"></a>'
+                '<span class="sortpriority" title="Sorting priority: 2">2</span>'
+                '<a href="?o=-2.-0.3" class="toggle ascending" title="Toggle sorting"></a>'
+                "</span>"
+            ),
+            html=True,
+        )
```
Comment (madmatah) [inline tests/ashley/machina_extensions/forum/test_forum_view.py:389]: Ok I assume that templates are tested here :slightly_smiling_face: 
You can ignore my comment on the fact that you test the response context, then ;)

**Review by madmatah — COMMENTED** [review]
Mostly comments on style, that's great @carofun ! :clap: :star_struck:

File: tests/ashley/machina_extensions/forum/test_forum_view.py (hunk 5 above)
Comment (carofun) [reply tests/ashley/machina_extensions/forum/test_forum_view.py:101]: I've build it this way because testing links in the HTML is heavy and doesn't make me validate the building process. I've made a test that validates that what I see in my template corresponds to what I have in my links which certify these tests.

File: tests/ashley/machina_extensions/forum/test_forum_view.py (hunk 6 above)
Comment (carofun) [reply tests/ashley/machina_extensions/forum/test_forum_view.py:328]: that is the test I'm talking about, it links my object with the what you see in the template and validates that the links that are generated are properly shown

File: tests/ashley/machina_extensions/forum/test_forum_view.py (hunk 7 above)
Comment (carofun) [reply tests/ashley/machina_extensions/forum/test_forum_view.py:389]: ok then great !

File: src/ashley/factories.py [hunk 8]
```diff
@@ -98,3 +98,22 @@ def content(self):
             "entityMap": {},
         }
         return json.dumps(data)
+
+
+class TopicSortHeaderFactory:
```
[the remark this answers is not in the document]
Comment (carofun) [reply src/ashley/factories.py:103]: Yes my idea was to build a factory class for tests, I thought the ones in the file had the same purpose. If it's not the case, a folder in test would be better

[the remark this answers is not in the document]
Comment (madmatah) [reply src/ashley/factories.py:103]: As I recall, some of these classes are used by the sandbox (for the fake LTI consumer), that's why it's not in the test package. But I think you can create a module in the test package to put this one.

I understand your need for this helper, but maybe a simple function might have been more relevant than a class with a single static method ?

[the remark this answers is not in the document]
Comment (carofun) [reply src/ashley/factories.py:103]: yes true, I can just put it in this test

File: tests/ashley/machina_extensions/forum/test_forum_view.py (hunk 6 above)
Comment (carofun) [reply tests/ashley/machina_extensions/forum/test_forum_view.py]: no problem madmatah :-p

File: tests/ashley/machina_extensions/forum/test_forum_view.py [hunk 9]
```diff
@@ -0,0 +1,411 @@
...
+
+from ashley.factories import (
+    ForumFactory,
+    PostFactory,
+    TopicFactory,
+    TopicSortHeaderFactory,
+    UserFactory,
+)
+
+logger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (carofun) [reply tests/ashley/machina_extensions/forum/test_forum_view.py:8]: It's used in the setUp method to create our Topics and Forum for the test

File: src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py (hunk 3 above)
Comment (madmatah) [reply src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py:35]: @sampaccoud To better understand, you suggest to submit this whole multi-sort feature to django-machina ? Or just the index part ?

Comment (carofun) [reply src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py:35]: Discussing this with Sam, we've decided to go ahead in our own repo and first validate this PR then do a PR on Django Machina which we hope will be accepted.
Then when we'll update our repo with the new version of Django machina containing the sorting, to get rid of the unnecessary code on our side. 
As Sam is quite sure it will get approved, his idea was not to override the model and add index in our repo because that could get messy once our PR gets accepted on Django machina

Comment (sampaccoud) [reply src/ashley/machina_extensions/forum_conversation/migrations/0014_auto_20210302_1614.py:35]: @madmatah don't you think it's an interesting feature for the original project?

File: src/ashley/templates/forum_conversation/topic_list.html [hunk 10]
```diff
@@ -9,16 +9,31 @@
     <div class="card topiclist">
       <div class="p-0 card-header">
         <div class="row m-0 px-3 py-2">
-          <div class="pl-0 col-md-8 col-sm-9 col-12 topic-name-col"><h3 class="m-0 card-title h5 text-dark">{% trans topic_list_title %}</h3></div>
-          <div class="col-md-1 d-none d-md-block text-center text-nowrap topic-count-col">{% trans "Replies" %}</div>
-          <div class="col-md-1 d-none d-md-block text-center text-nowrap topic-count-col">{% trans "Views" %}</div>
-          <div class="pr-0 col-md-2 col-sm-3 d-none d-sm-block text-nowrap topic-last-post-col">{% trans "Last post" %}</div>
+              <div class="pl-0 col-md-6 col-sm-9 col-12 topic-name-col sortable {{ header.0.class_attrib }}">
+                <h3 class="m-0 card-title h5 text-dark">
+                      <a href="{{ header.0.url_primary }}">{% trans topic_list_title %} </a>
```
Comment (sampaccoud) [inline src/ashley/templates/forum_conversation/topic_list.html:14]:
```suggestion
                      <a href="{{ header.0.url_primary }}">{% trans topic_list_title %}</a>
```

File: tests/ashley/machina_extensions/forum/test_forum_view.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (carofun) [reply tests/ashley/machina_extensions/forum/test_forum_view.py:8]: ahhhhh
ahhhh

**Review by madmatah — COMMENTED** [review]
After reviewing the code, I just tested the feature and have the following comments:

1. When I sort a column, I can't see the :arrow_down_small: icon like on your screenshot.
(error 404 on `/static/ashley/img/sorting-icons.svg`). I don't see this file on your commits, so I assume something is missing.

2. The default sorting order for the columns "Replies", "Views" and "Last post" is ascending and I don't think this is the best choice for the users. Who want to see the older or less active topic? It means that everyone will have to click 2 time on these columns.

3. I am confused by the behavior of multi-column sorting. Example : 
    - I click on the «Topics» column to sort them alphabetically
    - Yay, my topics are sorted ! :tada: 
    - Then, I would like to refine my sorting by replies count. So I click on «Replies»
    - The replies column became the first sorting criteria, I'm confused! I wanted it as a secondary criteria. This is not a natural behavior.

What do you think about these feedbacks ? I'm concerned about the user experience here.

**Comment by sampaccoud** [comment]
>        * The replies column became the first sorting criteria, I'm confused! I wanted it as a secondary criteria. This is not a natural behavior.
> 
I don't think people expect "second order" sorting on a website? 
The more basic and classical way to offer sorting would be a select box... is this what you're thinking of?

**Comment by carofun** [comment]
> After reviewing the code, I just tested the feature and have the following comments:

> 3. I am confused by the behavior of multi-column sorting. Example :
>    
>    * I click on the «Topics» column to sort them alphabetically
>    * Yay, my topics are sorted ! 🎉
>    * Then, I would like to refine my sorting by replies count. So I click on «Replies»
>    * The replies column became the first sorting criteria, I'm confused! I wanted it as a secondary criteria. This is not a natural behavior.
> 
This is the way it's done in the admin...I did find it a bit strange at first but didn't have a second thought on it as this was asked specifically this way.

**Comment by madmatah** [comment]
> > * The replies column became the first sorting criteria, I'm confused!
> >   I wanted it as a secondary criteria. This is not a natural behavior.
> 
> I don't think people expect "second order" sorting on a website?
> The more basic and classical way to offer sorting would be a select box... is this what you're thinking of?

Maybe my example was incomplete, the actual behavior is : 

- The replies column became the first sorting criteria **and the Topics columns became the second sorting criteria**, I'm confused! I wanted «replies» as a secondary criteria. This is not a natural behavior.

This whole PR (and issue #66) is about multi-column sorting, so my feedbacks were going in this direction.
With this goal, I don't understand your answer.

Beside that, I agree with you that most users want a single column sort feature!

**Comment by sampaccoud** [comment]
> This whole PR (and issue #66) is about multi-column sorting, so my feedbacks were going in this direction.
> With this goal, I don't understand your answer.

Yes the original issue sucks :scream: I don't know why I took django admin as example :man_shrugging: 
So sorry... Let me find a rock to hide.
